[RFC] Bundle flow versions instead of post-install downloading

zertosh opened the pull request:
### Why do this?

* Zero dependencies.
* No download = no firewall issues.
* Ability to check into VCS without leaving other platforms out.
* `post-install` scripts are security liability.
  - Some people disable them with the `ignore-scripts` option, so to install `flow-bin` you have to make an exception for all of `flow-bin`'s dependencies. Or, manually trigger it after installing.

### Considerations

Bundling Mac, Linux, and Windows in the same package, is only 4MB larger than the current distribution when installed using npm3. However, when installed with npm2, it is 4MB smaller.

```sh
# Current (npm2)

$ time npm2 install flow-bin
npm-2 install flow-bin  11.27s user 3.97s system 101% cpu 14.957 total
$ du -d 1 -h
23M     ./node_modules

# Current (npm3)

$ time npm3 install flow-bin
npm3 install flow-bin  7.34s user 1.68s system 34% cpu 26.131 total
$ du -d 1 -h
15M     ./node_modules

# Bundling (npm3)

$ npm pack
$ du -d 1 -h
6.2M     flow-bin-0.30.0.tgz
$ time npm3 install ~/src/flow-bin/flow-bin-0.30.0.tgz
npm3 install ~/src/flow-bin/flow-bin-0.30.0.tgz  1.28s user 0.31s system 123% cpu 1.290 total
$ du -d 1 -h
19M     ./node_modules
```

What's actually included?

```
274B ./node_modules/SHASUM256.txt
597B ./node_modules/cli.js
4.9M ./node_modules/flow-linux64-v0.30.0/flow
6.1M ./node_modules/flow-osx-v0.30.0/flow
8.4M ./node_modules/flow-win64-v0.30.0/flow.exe
768B ./node_modules/index.js
1.5K ./node_modules/license
740B ./node_modules/package.json
706B ./node_modules/readme.md
```

Installing the bundled version *should* also be considerably faster (on a fast connection). `npm install` spends a lot of time tracing dependencies and resolving versions. A single tarball installs much faster than a lot of small ones.

The currently distributed flow binaries ship with debug symbols. Stripping them before publishing puts the bundle size on-par with the current version:

```
# before
4.9M flow-linux64-v0.30.0/flow
6.1M flow-osx-v0.30.0/flow

# after
3.7M flow-linux64-v0.30.0/flow
3.7M flow-osx-v0.30.0/flow
```

### How would releases work?

Documentation is not included in this PR - that includes release instructions. I'll do that if the reception of this PR is positive. But I'm thinking something like this:

1. Bump the `package.json`, since that's where the flow version comes from.
2. Run `./prepare.sh && npm test`.
3. Maybe do a quick file inspection?
4. Commit the new `SHASUM256.txt` generated by `prepare.sh` and `package.json`. The `SHASUM256.txt` is for other tools that may want to verify the integrity of the bins (like Nuclide). May want to also have a signed version.
5. `npm publish`

Personally, I think bins should be checked into github, but this PR leaves them out because that's contentious.

Thoughts?

zertosh commented:
cc: @nmote @thejameskyle

nmote commented:
As long as there are no dependencies we can checksum the entire folder and then run flow-bin by default in Nuclide without a security risk.

Or, better yet, the Flow team could sign it and then we wouldn't have to update Nuclide's list of checksums each time.

See https://github.com/facebook/nuclide/issues/570

gabelevi commented:
This is an interesting PR! My initial instinct is to take this change. I think your "why do this" section is convincing. My thoughts, in no particular order:

* At the moment, the size of flow-bin before/after is close enough that it doesn't bother me. However, the size of the new flow-bin will grow as the number of binaries we distribute grows. However, I don't think that's going to be a problem for awhile. 32-bit is the only additional platform I can think of that we have even considered supporting.
* I've seen people use the `flow-bin/vendor/flow` path before. Heck, I've used it too. This will be a breaking change, but maybe that's for the best. People probably should be using node_modules/.bin/flow anyway. I suppose we could help mitigate by putting a script in vendor/flow and vendor/flow.exe that just delegates to cli.js
* I don't have any ideological opposition to signing the binaries. How hard would that be to do? I see codesign for OS X and a bunch of documentation for Windows. And...err...bsign for Linux?
* Travis doesn't seem happy - can you take a look?

nmote commented:
> I don't have any ideological opposition to signing the binaries. How hard would that be to do? I see codesign for OS X and a bunch of documentation for Windows. And...err...bsign for Linux?

I don't know, but I suspect we'll want to use [Node's crypto library](https://nodejs.org/api/crypto.html) to verify signatures in Nuclide. We'll definitely have to do some research on best practices and approach it carefully; I'm always leery of crypto since there are so many opportunities to accidentally make something insecure.

zertosh commented:
@gabelevi

* I added `vendor/flow` and `vendor/flow.exe` as js scripts that pass-through to `cli.js`.
* I switched to `make` - hope that's ok, my original script was kinda clunky.
* As for signing, you don't have to actually sign the binary, you sign `SHASUM256.txt`. That's what they do for node anyway https://github.com/nodejs/node#verifying-binaries.
* I added release instructions, there's no `npm version` step (looks like you did that before). There's only one source of truth for the flow version, the version in `package.json`. So you can't really bump the version before running `make`. Although I did leave it open to run `FLOW_VERSION=0.28.0 make`. I know you can use `pre-publish` or `pre-version`, but I'm not a fan of not being in control when you're packaging - but I can add this if you like.
* I have a lot of junk commits, if you're satisfied with the PR, let me know so I can squash before you merge :)

@nmote

In Nuclide, we'd have to keep the flow team's public key and use one of the PGP libraries to verify the authenticity of `SHASUM256.txt.asc`. If it passes, then we trust the SHAs and test the binary against it before executing. (I'm thinking this is overkill, we should chat about it).

gabelevi commented on vendor/flow.exe:
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env node
```
I don't think Windows has shebang support.

Anyway, Windows support is new enough that we can probably just delete this file and no one will notice

HeavyStorm replied:
It does have. NPM takes care of this.

gabelevi commented on vendor/flow:
```diff
@@ -1,3 +1,4 @@
+#!/usr/bin/env node
```
Can we delete install.js too?

gabelevi commented:
Ok, I think I'm going to take this PR! I think you've made a very compelling argument, and this looks good to me. I'll poke the rest of the team to make sure they don't have any objections and then merge. Thanks a ton for putting in this work!

yunxing commented:
Thanks! Bundling binaries would be pretty useful in [other projects](https://github.com/facebook/reason/issues/735) as well.

zertosh commented:
I'll address the last feedback sometime this weekend so we can merge this. I'm mid-cross-country move right now

yunxing commented on index.js:
```diff
@@ -7,6 +7,16 @@
-var bin = require('./lib');
 
-module.exports = bin.path();
+var VERSION = require('./package.json').version;
```
Like how you get the version number.

gabelevi commented:
I'm about to publish a new flow-bin, so I'm going to merge this and then tweak things myself :)

gabelevi commented:
Manually merged this as ec25746916e7619fef461e0ae79b40e6dd16f9d8 and just published v0.32.0, using this PR! It seems to work! Fingers crossed!

And thanks again @zertosh for building this!
